### PR TITLE
Fix n+1 problems

### PR DESCRIPTION
**Pull request by migo1**
### MILESTONES

- Write integration tests with Capybara gem.
- Secure app from n+1 problems.

**this is before fixing N+1 queries**
before
**After fixing N+1 queries**
after

### CODE REVIEW

File: spec/views/users_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+require 'rails_helper'
```
Comment (Lucas-Masaba) [inline spec/views/users_index_spec.rb:1]: - [ ] I ran your view tests locally and got failed tests as seen in the screenshot below.
image

[OPTIONAL]
- I ran all your tests in your spec folder and got the errors seen in the screenshot below.  Please fix these
image

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -3,7 +3,9 @@ class PostsController < ApplicationController
   before_action :set_post, only: [:show]
 
   def index
-    @posts = @user.posts
+    @posts = Post.includes(:author, :comments, :likes)
```
Comment (Lucas-Masaba) [inline app/controllers/posts_controller.rb]: - [x] In order to solve N+1 problems, you should add `.includes` to `@user.posts` you had previously. It should look something like this
```ruby
@posts = @user.posts.includes(:comments)
```

File: spec/views/users_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    user2 = User.create(name: 'User2',
+                        photo: 'https://free-url-shortener.rb.gy/url-shortener.png', posts_counter: 0)
+    Post.create(author: user1, title: 'first post', comments_counter: 0, likes_counter: 0)
+    Post.create(author: user2, title: 'second post', comments_counter: 0, likes_counter: 0)
+    Post.create(author: user2, title: 'third post', comments_counter: 0, likes_counter: 0)
+
+    visit users_path
+
+    expect(page).to have_content('1')
+    expect(page).to have_content('2')
```
Comment (Lucas-Masaba) [inline spec/views/users_index_spec.rb]: - [x] Write a test that asserts you can see the number of posts each user has written.

**Review by Lucas-Masaba — CHANGES_REQUESTED** [review]
# Hello @migo1 && @souad988  👋 , 

Good job so far! 👍 
There are some issues that you still need to work on to go to the next project but you are almost there!

almost-there

## Highlights 🍾 
- Linter checks passing :heavy_check_mark: 
- Well written pull request message and commit messages :writing_hand: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions 💭 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

### [OPTIONAL]
- Please add your tests in `describe` or `scenario` blocks because they are easier to read and track any issues with your tests.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @Lucas-Masaba in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/users_index_spec.rb (hunk 1 above)
Comment (migo1) [reply spec/views/users_index_spec.rb:1]: @Lucas-Masaba, thanks for the review

the test are all passing from my computer
Screenshot from 2023-10-03 20-20-37

File: spec/views/users_index_spec.rb (hunk 3 above)
Comment (migo1) [reply spec/views/users_index_spec.rb]: noted

File: spec/views/users_index_spec.rb (hunk 1 above)
Comment (Lucas-Masaba) [reply spec/views/users_index_spec.rb:1]: Hello @migo1 , if they are all passing on your machine, then that is alright. :+1:  But, could you please check that they are running on your coding partner's machine just to confirm?

File: app/controllers/posts_controller.rb (hunk 2 above)
Comment (migo1) [reply app/controllers/posts_controller.rb]: @Lucas-Masaba, noted

File: spec/views/users_index_spec.rb (hunk 1 above)
Comment (migo1) [reply spec/views/users_index_spec.rb:1]: @Lucas-Masaba Yes, just confirmed that they are passing

**Review by nedjwab — APPROVED** [review]
# 🟢 Project Approved! 🏆

## Hi  @team               👋

Your project is complete! ✔️✔  🏁

Your code was very professional 💯

There is nothing else to say other than... _it's time to merge it_ :shipit:
Congratulations! 🎉

 
## Highlights 🌟

- ✔️ Descriptive commmit messsages
- ✔️ Descriptive PR description
- ✔️ No linter Errors


- The code is top-notch! 💯 and have a huge potential.

Cheers and Happy coding!👏👏👏

## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.


_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**